src/game/client/draw/w3dmodeldraw.cpp

```diff
@@ -1154,3 +1178,1890 @@ void W3DModelDrawModuleData::Parse_Condition_State(INI *ini, void *instance, voi
+    if (m_curState == nullptr || !Maintain_Frame_Across_States(m_curState->m_flags) || !m_renderObject
+        || m_renderObject->Class_ID() != RenderObjClass::CLASSID_HLOD) {
+        return -1.0f;
+    }
+
+    HLodClass *hlod = static_cast<HLodClass *>(m_renderObject);
+
+    float frame;
+    int frames;
+    int mode;
+    float multiplier;
+    hlod->Peek_Animation_And_Info(frame, frames, mode, multiplier);
+
+    if (frame < 0.0f) {
+        return 0.0f;
+    }
+
+    if (frames > frame) {
+        return frame / (frames - 1.0f);
+    }
+
+    return 1.0f;
+}
+
+void W3DModelDraw::Adjust_Animation(ModelConditionInfo const *prev_state, float f)
+{
+    if (m_curState != nullptr) {
+        int count = m_curState->m_animations.size();
+
+        if (count <= 0) {
+            m_whichAnimInCurState = -1;
+        } else {
+            if (count == 1) {
+                m_whichAnimInCurState = 0;
+            } else if (prev_state == m_curState) {
+                int which = m_whichAnimInCurState;
+
+                while (m_whichAnimInCurState == which) {
+                    m_whichAnimInCurState = Get_Client_Random_Value(0, count - 1);
+                }
+            } else {
+                m_whichAnimInCurState = Get_Client_Random_Value(0, count - 1);
+            }
+
+            HAnimClass *anim = m_curState->m_animations[m_whichAnimInCurState].Get_Anim_Handle();
+
+            if (m_renderObject != nullptr && anim != nullptr) {
+                int frame = 0;
+
+                if (m_curState->m_mode == RenderObjClass::ANIM_MODE_ONCE_BACKWARDS
+                    || m_curState->m_mode == RenderObjClass::ANIM_MODE_LOOP_BACKWARDS) {
+                    frame = anim->Get_Num_Frames() - 1;
+                }
+
+                if (Test_Animation_Flag(m_curState->m_flags, RANDOMSTART)) {
+                    frame = Get_Client_Random_Value(0, anim->Get_Num_Frames() - 1);
+                } else if (Test_Animation_Flag(m_curState->m_flags, START_FRAME_FIRST)) {
+                    frame = 0;
+                } else if (Test_Animation_Flag(m_curState->m_flags, START_FRAME_LAST)) {
+                    frame = anim->Get_Num_Frames() - 1;
+                } else if (Maintain_Frame_Across_States(m_curState->m_flags) && prev_state != nullptr
+                    && prev_state != m_curState && Maintain_Frame_Across_States(prev_state->m_flags)
+                    && Maintain_Frame_Across_States_2(m_curState->m_flags, prev_state->m_flags) && f >= 0.0f) {
+                    frame = GameMath::Fast_To_Int_Truncate(anim->Get_Num_Frames() * f - 1.0f);
+                }
+
+                m_renderObject->Set_Animation(anim, frame, m_curState->m_mode);
+                anim->Release_Ref();
+
+                if (m_renderObject->Class_ID() == RenderObjClass::CLASSID_HLOD) {
+                    HLodClass *hlod = static_cast<HLodClass *>(m_renderObject);
+                    hlod->Set_Animation_Frame_Rate_Multiplier(Get_Client_Random_Value_Real(

```

Comment:
The line could be separated into two lines for readability, putting result of `Get_Client_Random_Value_Real` into variable first.